controllers/front/OnePageCheckoutController.php

```diff
@@ -0,0 +1,432 @@
+        $cart = $this->context->cart;
+
+        $data['checksum'] = $this->cartChecksum->generateChecksum($cart);
+
+        Db::getInstance()->execute(
+            'UPDATE ' . _DB_PREFIX_ . 'cart SET checkout_session_data = "' . pSQL(json_encode($data)) . '"
+                WHERE id_cart = ' . (int) $cart->id
+        );
+    }
+
+    /**
+     * Restores from checkout session some previously persisted cart-related data.
+     *
+     * @param CheckoutProcess $process
+     */
+    protected function restorePersistedData(CheckoutProcess $process)
+    {
+        $cart = $this->context->cart;
+        $customer = $this->context->customer;
+        $rawData = Db::getInstance()->getValue(
+            'SELECT checkout_session_data FROM ' . _DB_PREFIX_ . 'cart WHERE id_cart = ' . (int) $cart->id
+        );
+        $data = json_decode($rawData ?? '', true);
+        if (!is_array($data)) {
+            $data = [];
+        }
+
+        $addressValidator = new AddressValidator();
+        $invalidAddressIds = $addressValidator->validateCartAddresses($cart);
+
+        // Build the currently selected address' warning message (if relevant)
+        if (!$customer->isGuest() && !empty($invalidAddressIds)) {
+            $this->checkoutWarning['address'] = [
+                'id_address' => (int) reset($invalidAddressIds),
+                'exception' => $this->trans(
+                    'Your address is incomplete, please update it.',
+                    [],
+                    'Shop.Notifications.Error'
+                ),
+            ];
+        }
+
+        // Prevent check for guests
+        if ($customer->id) {
+            // Prepare all other addresses' warning messages (if relevant).
+            // These messages are displayed when changing the selected address.
+            $allInvalidAddressIds = $addressValidator->validateCustomerAddresses($customer, $this->context->language);
+            $this->checkoutWarning['invalid_addresses'] = $allInvalidAddressIds;
+        }
+
+        if (isset($data['checksum']) && $data['checksum'] === $this->cartChecksum->generateChecksum($cart)) {
+            $process->restorePersistedData($data);
+        }
+    }
+
+    public function displayAjaxselectDeliveryOption(): void
+    {
+        $cart = $this->cart_presenter->present(
+            $this->context->cart,
+            true
+        );
+
+        ob_end_clean();
+        header('Content-Type: application/json');
+        $this->ajaxRender(json_encode([
+            'preview' => $this->render('checkout/_partials/cart-summary', [
+                'cart' => $cart,
+                'static_token' => Tools::getToken(false),
+            ]),
+        ]));
+    }
+
+    public function displayAjaxCheckCartStillOrderable(): void
+    {
+        $responseData = [
+            'errors' => false,
+            'cartUrl' => '',
+        ];
+
+        if ($this->context->cart->isAllProductsInStock() !== true
```

Comment:
This should be centralized, it's already in the code like 3 times, every implementation checking something else. It would be good to have one source of truth called `isCartOrderable` that will check absolutely everyting, and call it in `CartController` and `OrderController`.